# Homework-3 

## Use the Titanic Dataset to Practice Data Wrangling Tasks
### Part 1: Import libaries , load the data, inspect the data, then get the name of the variables and create a small markdown table  with variable names and types (e.g numerical or categorical ordinal, categorical nominal etc.)

### Part 2: 
    - Impute missing values for "Embarked" with the mode (most frequent value)
    - Impute missing value for Age using two following strategies
        - A) Simple mean imputation (fill with the overal mean age)
        - B) Group-wise median imputation: for each combination of Sex and Pclass compute the median age. Impute missing age using the median age of passenger's sex and Pclass group

### Part 3: Create a FamilySize variable (FamilySize = SibSp + Parch +1) and compute:
    - Minimum, Maximum, and mean FamilySize
    - How many passengers were traveling alone
    
### Part 4:  Create a filtered dataframe of all passengers:
    - Who are female
    - Who are male 
    - Who are under 16 years old   
    - Compute the survival rate for the above filtered data
    - Compute survival rate by Sex and Pclass (use groupby and mean) and present the result in a markedown table (rows: sex, columns: Pclass) or using print formating.
    - Interpret your result in 2-3 sentences
     
    
### Part 5: Compute the average fare for each Pclass (Fare values are per tickets not per person (there coud be more than 1 person on a ticket) so make sure to account for that. 

## Part 1

In [2]:
import pandas as pd
import numpy as np

df = pd.read_csv("C:/Users/hwang/OneDrive/Documents/MC stuff/Spring 2026/DATA 101 Introduction to Data Science/Classwork/Datasets/titanic.csv")

In [3]:
print(df.head())

   PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  


In [4]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB
None


In [5]:
print(df.columns.tolist())

['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked']


Variable name | Type

PassengerId    Categorical (Nominal)

Survived	   Categorical (Nominal)

Pclass	       Categorical (Ordinal)

Name           Categorical (Nominal)

Sex	           Categorical (Nominal)

Age	           Quantitative/Numerical

SibSp          Quantitative/Numerical

Parch          Quantitative/Numerical

Ticket         Categorical (Nominal)

Fare           Quantitative/Numerical

Cabin          Categorical (Nominal)

Embarked	   Categorical (Nominal)

# Part 2

In [11]:
embarked_mode = df['Embarked'].mode()[0]

df['Embarked'] = df['Embarked'].fillna(embarked_mode)

print(f"Imputed Embarked with: {embarked_mode}")

Imputed Embarked with: S


In [39]:
mean_age = df_mean['Age'].mean()

df_mean['Age'] = df_mean['Age'].fillna(mean_age)

print(f"Overall Mean Age: {mean_age: }")

Overall Mean Age:  29.69911764705882


In [46]:
df['Age'] = df['Age'].fillna(df.groupby(['Sex', 'Pclass'])['Age'].transform('median'))

print("Missing values in Age after Strategy B:", df['Age'].isnull().sum())

Missing values in Age after Strategy B: 0


# Part 3

In [33]:
df['FamilySize'] = df['SibSp'] + df['Parch'] + 1

print(df[['SibSp', 'Parch', 'FamilySize']].head())

   SibSp  Parch  FamilySize
0      1      0           2
1      1      0           2
2      0      0           1
3      1      0           2
4      0      0           1


In [40]:
family_stats = df['FamilySize'].agg(['min', 'max', 'mean'])

print(f"Minimum Family Size: {family_stats['min']}")
print(f"Maximum Family Size: {family_stats['max']}")
print(f"Mean Family Size:    {family_stats['mean']: }")

Minimum Family Size: 1.0
Maximum Family Size: 11.0
Mean Family Size:     1.904601571268238


In [47]:
passengers_alone = (df['FamilySize'] == 1).sum()

print(f"Number of passengers traveling alone: {passengers_alone}")

percent_alone = (passengers_alone / len(df)) * 100
print(f"Percentage of alone passengers: {percent_alone: }%")

Number of passengers traveling alone: 537
Percentage of alone passengers:  60.26936026936027%


# Part 4

In [48]:
female_survival = df[df['Sex'] == 'female']['Survived'].mean()
male_survival = df[df['Sex'] == 'male']['Survived'].mean()
under_16_survival = df[df['Age'] < 16]['Survived'].mean()

print(f"Survival Rate for Females: {female_survival: %}")
print(f"Survival Rate for Males:   {male_survival: %}")
print(f"Survival Rate for Under 16: {under_16_survival: %}")

Survival Rate for Females:  74.203822%
Survival Rate for Males:    18.890815%
Survival Rate for Under 16:  59.036145%


In [50]:
survival_table = df.groupby(['Sex', 'Pclass'])['Survived'].mean()

print(survival_table)

Sex     Pclass
female  1         0.968085
        2         0.921053
        3         0.500000
male    1         0.368852
        2         0.157407
        3         0.135447
Name: Survived, dtype: float64


#### The data confirms the "Women and Children First" rule, as women and children under 16 had significantly higher survival rates than adult males. However, social class played a massive role. A woman in 1st class had over a 45% chance difference of surviving compared to a woman in 3rd class. Ultimately, the highest chance of survival belonged to 1st-class females (almost 97%), while the lowest was reserved for 3rd class males (approximately 13.5%).

# Part 5

In [45]:
df['GroupSize'] = df.groupby('Ticket')['Ticket'].transform('count')

df['FarePerPerson'] = df['Fare'] / df['GroupSize']

avg_fare_per_class = df.groupby('Pclass')['FarePerPerson'].mean()

print("Average Fare Per Person by Class:")
print(avg_fare_per_class)

Average Fare Per Person by Class:
Pclass
1    43.650347
2    13.322599
3     8.085857
Name: FarePerPerson, dtype: float64
